### PR TITLE
added binary_search and added README note

### PR DESCRIPTION
**Pull request by radicalrafi**
#### Short description of what this resolves:
This implement binary search it's a relatively easy to implement ,I suggest implementations start using generic paradigm (templates) instead of typed ones, I also suggest to use std::vector instead of C-like arrays for performance,memory safety and iterators .
Hacktoberfest :+1: 
#### Changes proposed in this pull request:

- ...
- ...

**Fixes**: #

---

Before submitting, pleqase check if your submission conforms with the following checklist:
- [ ] Add tests for all implementations.
- [x] Descriptive pull request message
- [x] Descriptive commit messages
- [x] write test cases as written in the test folder
- [ ] use a formatter to format your code according to the google standards.
- [x] folder names should contain underscores '_' instead of spaces
- [ ] please write full and descriptive names for variables
- [x] remove main function from the code and add the tests in the tests folder by creating a new file in a similar structure and similar name as others.
- [x] keep your implementation to C++, and write the code in .hpp files
- [x] tests are written in .cpp files

### CODE REVIEW

File: src/search/binary.hpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#include <bits/stdc++.h>
```
Comment (manparvesh) [inline src/search/binary.hpp]: I don't think we need to import this library. Just import `algorithm`

File: README.md [hunk 2]
```diff
@@ -53,6 +53,12 @@ $ sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
 $ sudo apt-get update -qq
 $ sudo apt-get install g++-6
 $ sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-6 90
+
+
+- Run in a Terminal
+```
+git submodule update --init --recursive
+```
```
Comment (manparvesh) [inline README.md]: This is not needed. I have already written in the instructions to clone with the `--recursive` flag

File: src/search/binary.hpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    /*
+        for optimal performance the array needs to be sorted so we run quick sort 
+        thus we make sure the array is sorted without having to check if it's sorted
+        
+        if we wanted to check  we can use is_sorted for algorithm header but it requires passing an array
+        that accepts std::vector or std::array not C-like arrats
+    */
+    quick_sort(n,a);
+    auto L = 0;
+    auto R = n -1;
```
Comment (manparvesh) [inline src/search/binary.hpp]: use lowercase alphabets for variable names.

File: src/search/binary.hpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    while (L < R){
+        float _x = (L+R)/2;
```
Comment (manparvesh) [inline src/search/binary.hpp]: use descriptive names instead of just anything.

File: src/search/binary.hpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        auto m = floor(_x);
+        if (a[m] < T){
+            L = m+1;
```
Comment (manparvesh) [inline src/search/binary.hpp]: reformat the whole file please. Use a formatter.

File: test/search/test_binary.cpp [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#include "gtest/gtest.h"
+#include "binary.hpp"
+#include <bits/stdc++.h>
+using namespace std;
+
+TEST(insertion_sort, all_tests){
```
Comment (manparvesh) [inline test/search/test_binary.cpp]: the name of test is `binary_search`, not `insertion_sort`

File: test/search/test_binary.cpp [hunk 7]
```diff
@@ -0,0 +1,10 @@
+#include "gtest/gtest.h"
+#include "binary.hpp"
+#include <bits/stdc++.h>
```
Comment (manparvesh) [inline test/search/test_binary.cpp]: remove this line.

**Comment by radicalrafi** [comment]
Changes done submitting another pr

**Pull request closed without merging**